src/puppetlabs/puppetdb/amq_migration.clj

```diff
@@ -136,31 +137,36 @@
           (finally
             (.acknowledge message)))))))
 
-(defn consume-everything
-  "Attempts to read all available messages from ActiveMQ, giving up
-  after 5 seconds if one is not available"
-  [^MessageConsumer consumer process-message id]
-  (try
-    (loop []
-      (when-let [msg (.receive consumer 5000)]
-        (process-message (swap! id inc) msg)
-        (recur)))
-    (catch javax.jms.IllegalStateException e
-      (log/info (i18n/trs "Received IllegalStateException, shutting down")))
-    (catch Exception e
-      (log/error e))))
-
 (defn transfer-mq-messages
+  "Returns true if it's likely that all of the existing messages were
+  transferred."
   [^ConnectionFactory conn-pool endpoint process-message id]
   (with-open [connection (create-connection conn-pool)
               session (create-session connection)
+              producer (.createProducer session (create-queue session endpoint))
               consumer (.createConsumer session (create-queue session endpoint))]
     (log/info (i18n/trs "Processing messages from the Queue: {0}" endpoint))
-    (consume-everything consumer process-message id)))
+    (mq/send-message! connection endpoint
+                      "\"ignore me\""
+                      {"command" "end-of-line"})
+    (try
+      (loop []
+        (when-let [msg (.receive consumer 120000)]
```

Comment:
I think so, not sure what the best way to go about it is. We wait for 2 minutes, but 2 minutes would be very abnormal. I would expect normal situations to get a message immediately, certainly within a few seconds. Maybe we can make one receive call with something like a 10 second wait. If we get nothing, we log that we're going to wait for 2 more minutes (maybe warn?) and invoke receive again? What do you think @wkalt ?